draft-packages/popover/KaizenDraft/Popover/PopoverModern.tsx

```diff
@@ -0,0 +1,179 @@
+  automationId,
+  children,
+  variant = "default",
+  placement = "bottom",
+  size = "small",
+  heading,
+  dismissible = false,
+  onClose,
+  singleLine = false,
+  customIcon,
+  referenceElement,
+}: ModernPopoverProps) => {
+  const [popperElement, setPopperElement] = useState<HTMLDivElement | null>(
+    null
+  )
+  const [arrowElement, setArrowElement] = useState<HTMLDivElement | null>(null)
+  const { styles: popperStyles, attributes } = usePopper(
+    referenceElement,
+    popperElement,
+    {
+      modifiers: [
+        {
+          name: "arrow",
+          options: {
+            element: arrowElement,
+            // Ensures that the arrow doesn't go too far to the left or right
+            // of the tooltip.
+            padding: arrowWidth / 2 + 10,
+          },
+        },
+        {
+          name: "offset",
+          options: {
+            offset: [0, arrowHeight],
+          },
+        },
+        {
+          name: "preventOverflow",
+          options: {
+            // Makes sure that the popover isn't flush up against the end of the
+            // viewport
+            padding: 4,
+          },
+        },
+      ],
+      placement,
+    }
+  )
+
+  return (
+    <div
+      ref={setPopperElement}
+      style={popperStyles.popper}
+      {...attributes.popper}
+      className={classNames(styles.root, mapSizeToClass(size))}
+      data-automation-id={automationId}
+    >
+      <div className={mapVariantToBoxClass(variant)}>
+        {heading && (
+          <div className={styles.header}>
+            {variant !== "default" && (
+              <span
+                className={classNames(
+                  styles.icon,
+                  mapVariantToIconClass(variant)
+                )}
+              >
+                <Icon
+                  role="presentation"
+                  icon={customIcon ? customIcon : mapVariantToIcon(variant)}
+                />
+              </span>
+            )}
+            <div className={styles.singleLine}>{heading}</div>
+            {dismissible && (
+              <button className={styles.close} onClick={onClose}>
+                <Icon role="presentation" icon={closeIcon} />
+              </button>
+            )}
+          </div>
+        )}
+        <div
+          className={classNames(styles.container, mapLineVariant(singleLine))}
+        >
+          {children}
+        </div>
+      </div>
+      <div
+        ref={setArrowElement}
+        style={popperStyles.arrow}
+        className={styles.arrowWrapper}
+      >
+        <div
+          className={classNames(styles.arrow, mapArrowVariantToClass(variant))}
+        />
+      </div>
+    </div>
+  )
+}
+
+/**
+ * How to use:
+ *
+ * const [referenceElement, referenceElementRef] = usePopoverReferenceElementRef()
+ *
+ * return (<>
+ *   <button ref={referenceElementRef}>
+ *     Hello world
+ *   </button>
+ *   <Popover referenceElement={referenceElement} />
+ * </>)
+ *
```

Comment:
This API feels a bit awkward to me. It’s not super difficult to understand, but it feels like most of this work could be abstracted away from users in _most_ use cases (though you might want to leave the basic API of `<Popper>` intact.

I can imagine something like this:

```tsx
const [popoverReferenceRef, PopoverWithCurriedRef] = usePopover();

return (
  <>
    <button ref={popoverReferenceRef}>Hello, World!</button>
    <PopoverWithCurriedRef>
      <p>Content</p>
    <PopoverWithCurriedRef>
  </>
)
```

With `usePopover` being something vaguely like:

```tsx
function usePopover() {
  const [refElement, setRefElement] = useState<HTMLElement | null>(null)
  const [popoverComponent, setPopoverComponent] = useState<React.FunctionComponent>(() => undefined)
  const ref = useCallback(node: HTMLElement => {
    setRefElement(node)
  })

  useEffect(() => {
    if (refElement) {
      setPopoverComponent((props) => {
        return <Popover {...props} referenceElement={refElement}/>
      })
    }
  }, [refElement, setPopoverComponent])

  return [ref, popoverComponent]
}
```

That _feels_ like a more natural API to me, since consumers don’t have to worry about handling nearly as much passing variables around. WDYR?

Reply:
Yeah, that would be easier to understand as a consumer. I've updated this now, LMK if all good. Only difference, I used `useMemo` instead of `useEffect`/`useState`, to avoid a double render.